lib/internal/process/coverage.js

```diff
@@ -0,0 +1,69 @@
+'use strict';
+const path = require('path');
+const { mkdirSync, writeFileSync } = require('fs');
+const inspector = require('inspector');
+// TODO(addaleax): add support for coverage to worker threads.
+const hasInspector = process.config.variables.v8_enable_inspector === 1 &&
+  require('internal/worker').isMainThread;
+
+let session;
+
+function writeCoverage() {
+  if (!hasInspector || !session) {
+    return;
+  }
+
+  const filename = `coverage-${process.pid}-${Date.now()}.json`;
+  try {
+    // TODO(bcoe): switch to mkdirp once #22302 is addressed.
+    mkdirSync(process.env.NODE_V8_COVERAGE);
+  } catch (err) {
+    if (err.code !== 'EEXIST') {
+      console.error(err);
+      return;
+    }
+  }
+
+  const target = path.join(process.env.NODE_V8_COVERAGE, filename);
+
+  try {
+    session.post('Profiler.takePreciseCoverage', (err, coverageInfo) => {
+      try {
+        if (err) throw err;
+        writeFileSync(target, JSON.stringify(coverageInfo));
+      } catch (err) {
+        console.error(err);
+      }
+    });
+  } catch (err) {
+    console.error(err);
+  } finally {
+    session.disconnect();
+    session = null;
+  }
+}
+
+exports.writeCoverage = writeCoverage;
+
+function setup() {
+  if (!hasInspector) {
+    return;
```

Comment:
I suggest to throw in error in this case. Otherwise the user would not know that no data will be collected.

In that case the `hasInspector` check in `writeCoverage` would also be obsolete.

Reply:
I’m not a fan of throwing here. If necessary, we can print a warning, but in practice, this situation really only occurs within Workers (for now).

Reply:
I've added a warning, which I feel is a reasonable compromise; It would be miserable to be unable to run your test suite with coverage because executing code in a worker results in exceptions being thrown.

When we support workers in the future, perhaps we could switch to throwing an exception?